issue: "module not found" when importing zodResolver under Zod v3

### Version Number

5.2.1

### Codesandbox/Expo snack

https://codesandbox.io/p/devbox/romantic-burnell-vsq3yy

### Steps to reproduce

1. Install `@hookform/resolvers@5.2.1` and `zod@3.24.1`
2. Import `zodResolver`

The import fails because it depends on `zod/v4/core`, which [was added in zod@3.25.0](https://github.com/colinhacks/zod/releases/tag/v3.25).

### Expected behaviour

I'd expect this kind of change to get a major version bump and a note in the release notes. Instead the release notes just say "[retains compatibility with Zod v3](https://github.com/react-hook-form/resolvers/releases/tag/v5.1.0)".

The [pull request](https://github.com/react-hook-form/resolvers/pull/777) that added the `zod/v4/core` import does mention this:

> This PR bumps the minimum Zod version to 3.25.0…

But AFAICT that's just referring to the `devDependencies` bump. The PR also says:

> I've designed this to continue working with Zod versions pre-3.25.0, but it's on a best-effort basis.

Maybe I'm missing something, but I'm not seeing how this could continue working with Zod versions pre-3.25.0…

Anyway, I spent some time scratching my head on this one since I'd wanted to upgrade my `@hookform/resolvers` dependency first before attempting a Zod v4 upgrade. I was able to resolve the problem once I learned that Zod 3.25.0 added the `zod/v4/core` path. I just wanted to file a ticket to save any other devs hitting this some time.

### What browsers are you seeing the problem on?

_No response_

### Relevant log output

```shell
✘ [ERROR] Missing "./v4/core" specifier in "zod" package [plugin vite:dep-pre-bundle]

    node_modules/.pnpm/@hookform+resolvers@5.2.1_react-hook-form@7.4.1/node_modules/@hookform/resolvers/zod/dist/zod.mjs:1:142:
      1 │ ...{appendErrors as o}from"react-hook-form";import*as n from"zod/v4/core";function t(r,e){try{var o=r()}catch(r){return e(r)}...
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

--- a/zod/src/__tests__/subpath-imports.ts
+++ b/zod/src/__tests__/subpath-imports.ts
@@ -1,0 +1,16 @@
+import { readFileSync } from 'fs';
+import { join } from 'path';
+
+// `zod/v3` and `zod/v4/core` don't exist before zod@3.25.0. Importing either
+// as a value (rather than `import type`) makes the compiled output reference
+// that subpath at runtime, which breaks module resolution for anyone on an
+// older zod 3.x — even if they never touch the zod4 code path (#811).
+test('zod/v3 and zod/v4/core are only ever imported as types', () => {
+  const source = readFileSync(join(__dirname, '..', 'zod.ts'), 'utf8');
+
+  const v3Import = source.match(/^import .*'zod\/v3';$/m)?.[0];
+  const v4Import = source.match(/^import .*'zod\/v4\/core';$/m)?.[0];
+
+  expect(v3Import).toMatch(/^import type /);
+  expect(v4Import).toMatch(/^import type /);
+});

--- a/zod/src/zod.ts
+++ b/zod/src/zod.ts
@@ -7,8 +7,13 @@ import {
   ResolverSuccess,
   appendErrors,
 } from 'react-hook-form';
-import * as z3 from 'zod/v3';
-import * as z4 from 'zod/v4/core';
+// Type-only: these subpaths don't exist before zod@3.25.0, so importing them
+// as values would break resolution for anyone on an older zod 3.x. `import
+// type` is fully erased at build time, so the compiled output never
+// references these subpaths at runtime — see `isZod4Error` below for how the
+// one runtime check that used to need `z4.$ZodError` avoids that now.
+import type * as z3 from 'zod/v3';
+import type * as z4 from 'zod/v4/core';
 
 const isZod3Error = (error: any): error is z3.ZodError => {
   return Array.isArray(error?.issues);
@@ -21,8 +26,12 @@ const isZod3Schema = (schema: any): schema is z3.ZodSchema => {
   return '_def' in schema && typeof schema._def === 'object';
 };
 const isZod4Error = (error: any): error is z4.$ZodError => {
-  // instanceof is safe in Zod 4 (uses Symbol.hasInstance)
-  return error instanceof z4.$ZodError;
+  // Replicates zod4's own `$ZodError`/`Symbol.hasInstance` check
+  // (`inst?._zod?.traits?.has(name)`, see zod's `$constructor` in
+  // `v4/core/core.ts`) by hand rather than importing `$ZodError` as a value,
+  // since that import only exists in zod@3.25.0+ (see the `import type` note
+  // above).
+  return !!error?._zod?.traits?.has('$ZodError');
 };
 const isZod4Schema = (schema: any): schema is z4.$ZodType => {
   return '_zod' in schema && typeof schema._zod === 'object';

